Error when try to build to a folder that doesn't exist

**V version:**  `V 0.1.28 e666209`
**OS:**  macOS Catalina 10.15

**What did you do?**

Tried to build to the `bin` folder that didn't exist
```
 $ v -o bin/main main.v
```


**What did you expect to see?**

Something like:
```
 $ v -o bin/main main.v
yo `bin` folder doesn't exist
```

**What did you see instead?**
```
==================
ld: can't open output file for writing: bin/main, errno=2 for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

If you were not working with C interop, please raise an issue on GitHub:

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

make: *** [build] Error 1
```

--- a/vlib/os/os.v
+++ b/vlib/os/os.v
@@ -519,11 +519,27 @@ pub fn is_writable_folder(folder string) ?bool {
 	if !os.is_dir(folder) {
 		return error('`folder` is not a folder')
 	}
-	tmp_perm_check := os.join_path(folder, 'tmp_perm_check')
-	mut f := os.open_file(tmp_perm_check, 'w+', 0o700) or {
-		return error('cannot write to folder `$folder`: $err')
+	tmp_perm_check := os.join_path(folder, 'XXXXXX')
+	$if windows {
+		unsafe {
+			x := C._mktemp_s( tmp_perm_check.str, tmp_perm_check.len )
+			if 0 != x {
+				return error('C._mktemp_s failed')
+			}
+		}
+		mut f := os.open_file(tmp_perm_check, 'w+', 0o700) or {
+			return error('cannot write to folder $folder: $err')
+		}
+		f.close()
+	} $else {
+		unsafe {
+			x := C.mkstemp(tmp_perm_check.str)
+			if -1 == x {
+				return error('folder `$folder` is not writable')
+			}
+			C.close(x)
+		}
 	}
-	f.close()
 	os.rm(tmp_perm_check)
 	return true
 }

--- a/vlib/os/os.v
+++ b/vlib/os/os.v
@@ -511,39 +511,6 @@ pub fn is_executable(path string) bool {
   return C.access(charptr(path.str), x_ok) != -1
 }
 
-// `is_writable_folder` - `folder` exists and is writable to the process
-pub fn is_writable_folder(folder string) ?bool {
-	if !os.exists(folder) {
-		return error('`$folder` does not exist')
-	}
-	if !os.is_dir(folder) {
-		return error('`folder` is not a folder')
-	}
-	tmp_perm_check := os.join_path(folder, 'XXXXXX')
-	$if windows {
-		unsafe {
-			x := C._mktemp_s( tmp_perm_check.str, tmp_perm_check.len )
-			if 0 != x {
-				return error('C._mktemp_s failed')
-			}
-		}
-		mut f := os.open_file(tmp_perm_check, 'w+', 0o700) or {
-			return error('cannot write to folder $folder: $err')
-		}
-		f.close()
-	} $else {
-		unsafe {
-			x := C.mkstemp(tmp_perm_check.str)
-			if -1 == x {
-				return error('folder `$folder` is not writable')
-			}
-			C.close(x)
-		}
-	}
-	os.rm(tmp_perm_check)
-	return true
-}
-
 // `is_writable` returns `true` if `path` is writable.
 pub fn is_writable(path string) bool {
   $if windows {

--- a/vlib/os/os.v
+++ b/vlib/os/os.v
@@ -638,7 +638,7 @@ pub fn dir(path string) string {
 
 pub fn base_dir(path string) string {
 	posx := path.last_index(path_separator) or {
-		return path
+		return path.clone()
 	}
 	// NB: *without* terminating /
 	return path[..posx]

--- a/vlib/os/os_nix.c.v
+++ b/vlib/os/os_nix.c.v
@@ -28,8 +28,6 @@ mut:
 
 fn C.uname(name voidptr) int
 
-fn C.mkstemp(stemplate byteptr) int
-
 fn C.symlink(arg_1, arg_2 charptr) int
 
 pub fn uname() Uname {
@@ -191,4 +189,25 @@ pub fn (mut f File) close() {
 
 pub fn debugger_present() bool {
 	return false
+}
+
+fn C.mkstemp(stemplate byteptr) int
+// `is_writable_folder` - `folder` exists and is writable to the process
+pub fn is_writable_folder(folder string) ?bool {
+	if !os.exists(folder) {
+		return error('`$folder` does not exist')
+	}
+	if !os.is_dir(folder) {
+		return error('`folder` is not a folder')
+	}
+	tmp_perm_check := os.join_path(folder, 'XXXXXX')
+	unsafe {
+		x := C.mkstemp(tmp_perm_check.str)
+		if -1 == x {
+			return error('folder `$folder` is not writable')
+		}
+		C.close(x)
+	}
+	os.rm(tmp_perm_check)
+	return true
 }

--- a/vlib/os/os_nix.c.v
+++ b/vlib/os/os_nix.c.v
@@ -28,6 +28,8 @@ mut:
 
 fn C.uname(name voidptr) int
 
+fn C.mkstemp(stemplate byteptr) int
+
 fn C.symlink(arg_1, arg_2 charptr) int
 
 pub fn uname() Uname {

--- a/vlib/os/os_windows.c.v
+++ b/vlib/os/os_windows.c.v
@@ -294,7 +294,7 @@ pub fn exec(cmd string) ?Result {
 	}
 }
 
-fn C._mktemp_s(stemplate byteptr, stemplate_len int)
+fn C._mktemp_s(stemplate byteptr, stemplate_len int) int
 
 fn C.CreateSymbolicLinkW(&u16, &u16, u32) int
 

--- a/vlib/os/os_windows.c.v
+++ b/vlib/os/os_windows.c.v
@@ -294,8 +294,6 @@ pub fn exec(cmd string) ?Result {
 	}
 }
 
-fn C._mktemp_s(stemplate byteptr, stemplate_len int) int
-
 fn C.CreateSymbolicLinkW(&u16, &u16, u32) int
 
 pub fn symlink(origin, target string) ?bool {
@@ -367,4 +365,29 @@ pub fn uname() Uname {
 		version: unknown
 		machine: unknown
 	}
+}
+
+
+fn C._mktemp_s(stemplate byteptr, stemplate_len int) int
+// `is_writable_folder` - `folder` exists and is writable to the process
+pub fn is_writable_folder(folder string) ?bool {
+	if !os.exists(folder) {
+		return error('`$folder` does not exist')
+	}
+	if !os.is_dir(folder) {
+		return error('`folder` is not a folder')
+	}
+	tmp_perm_check := os.join_path(folder, 'XXXXXX')
+	unsafe {
+		x := C._mktemp_s( tmp_perm_check.str, tmp_perm_check.len )
+		if 0 != x {
+			return error('C._mktemp_s failed')
+		}
+	}
+	mut f := os.open_file(tmp_perm_check, 'w+', 0o700) or {
+		return error('cannot write to folder $folder: $err')
+	}
+	f.close()
+	os.rm(tmp_perm_check)
+	return true
 }

--- a/vlib/os/os_windows.c.v
+++ b/vlib/os/os_windows.c.v
@@ -294,6 +294,8 @@ pub fn exec(cmd string) ?Result {
 	}
 }
 
+fn C._mktemp_s(stemplate byteptr, stemplate_len int)
+
 fn C.CreateSymbolicLinkW(&u16, &u16, u32) int
 
 pub fn symlink(origin, target string) ?bool {

--- a/vlib/os/os_windows.c.v
+++ b/vlib/os/os_windows.c.v
@@ -368,7 +368,6 @@ pub fn uname() Uname {
 }
 
 
-fn C._mktemp_s(stemplate byteptr, stemplate_len int) int
 // `is_writable_folder` - `folder` exists and is writable to the process
 pub fn is_writable_folder(folder string) ?bool {
 	if !os.exists(folder) {
@@ -377,13 +376,7 @@ pub fn is_writable_folder(folder string) ?bool {
 	if !os.is_dir(folder) {
 		return error('`folder` is not a folder')
 	}
-	tmp_perm_check := os.join_path(folder, 'XXXXXX')
-	unsafe {
-		x := C._mktemp_s( tmp_perm_check.str, tmp_perm_check.len )
-		if 0 != x {
-			return error('C._mktemp_s failed')
-		}
-	}
+	tmp_perm_check := os.join_path(folder, 'tmp_perm_check')
 	mut f := os.open_file(tmp_perm_check, 'w+', 0o700) or {
 		return error('cannot write to folder $folder: $err')
 	}

--- a/vlib/v/builder/compile.v
+++ b/vlib/v/builder/compile.v
@@ -27,13 +27,15 @@ fn get_vtmp_filename(base_file_name, postfix string) string {
 }
 
 pub fn compile(command string, pref &pref.Preferences) {
-	// Check if the output directory is writable
-	output_folder := os.base_dir(pref.out_name)
-	os.is_writable_folder(output_folder) or {
-		println("Path `${output_folder}` is not writable or does not exist`")
-		return
+	odir := os.base_dir(pref.out_name)
+	// When pref.out_name is just the name of an executable, i.e. `./v -o executable main.v`
+	// without a folder component, just use the current folder instead:
+	output_folder := if odir.len == pref.out_name.len { os.getwd() } else { odir }
+	os.is_writable_folder(output_folder) or { 
+		// An early error here, is better than an unclear C error later:
+		verror(err)
+		exit(1)
 	}
-
 	// Construct the V object from command line arguments
 	mut b := new_builder(pref)
 	if pref.is_verbose {

--- a/vlib/v/builder/compile.v
+++ b/vlib/v/builder/compile.v
@@ -34,11 +34,10 @@ pub fn compile(command string, pref &pref.Preferences) {
 	if odir.len == pref.out_name.len {
 		output_folder = os.getwd()
 	}        
-	if !os.exists(output_folder) {
-		verror('folder $output_folder/ does not exist')
-	}
-	if !os.is_writable(output_folder) {
-		verror('folder $output_folder/ is not writable')
+	os.is_writable_folder(output_folder) or { 
+		// An early error here, is better than an unclear C error later:
+		verror(err)
+		exit(1)
 	}
 	// Construct the V object from command line arguments
 	mut b := new_builder(pref)

--- a/vlib/v/builder/compile.v
+++ b/vlib/v/builder/compile.v
@@ -34,10 +34,11 @@ pub fn compile(command string, pref &pref.Preferences) {
 	if odir.len == pref.out_name.len {
 		output_folder = os.getwd()
 	}        
-	os.is_writable_folder(output_folder) or { 
-		// An early error here, is better than an unclear C error later:
-		verror(err)
-		exit(1)
+	if !os.exists(output_folder) {
+		verror('folder $output_folder/ does not exist')
+	}
+	if !os.is_writable(output_folder) {
+		verror('folder $output_folder/ is not writable')
 	}
 	// Construct the V object from command line arguments
 	mut b := new_builder(pref)

--- a/vlib/v/builder/compile.v
+++ b/vlib/v/builder/compile.v
@@ -30,7 +30,10 @@ pub fn compile(command string, pref &pref.Preferences) {
 	odir := os.base_dir(pref.out_name)
 	// When pref.out_name is just the name of an executable, i.e. `./v -o executable main.v`
 	// without a folder component, just use the current folder instead:
-	output_folder := if odir.len == pref.out_name.len { os.getwd() } else { odir }
+	mut output_folder := odir
+	if odir.len == pref.out_name.len {
+		output_folder = os.getwd()
+	}        
 	os.is_writable_folder(output_folder) or { 
 		// An early error here, is better than an unclear C error later:
 		verror(err)

--- a/vlib/v/builder/compile.v
+++ b/vlib/v/builder/compile.v
@@ -27,6 +27,13 @@ fn get_vtmp_filename(base_file_name, postfix string) string {
 }
 
 pub fn compile(command string, pref &pref.Preferences) {
+	// Check if the output directory is writable
+	output_folder := os.base_dir(pref.out_name)
+	os.is_writable_folder(output_folder) or {
+		println("Path `${output_folder}` is not writable or does not exist`")
+		return
+	}
+
 	// Construct the V object from command line arguments
 	mut b := new_builder(pref)
 	if pref.is_verbose {

